src/provider.rs

```diff
@@ -172,6 +197,199 @@ impl TableProvider for SeafowlTable {
+                Err(error) => Err(DataFusionError::Internal(format!(
+                    "Failed to parse min/max value: {}",
+                    error
+                ))),
+            },
+            None => {
+                // Try to cast the None value to the corresponding ScalarValue matching `data_type`,
+                // e.g. `ScalarValue::Int16(None)`, `ScalarValue::Boolean(None)`, etc.
+                let value = ScalarValue::Null;
+                let cast_options = CastOptions { safe: false };
+                let cast_arr =
+                    cast_with_options(&value.to_array(), data_type, &cast_options)?;
+                Ok(ScalarValue::try_from_array(&cast_arr, 0)?)
+            }
+        }
+    }
+
+    // Prune away partitions that are refuted by the provided filter expressions
+    async fn prune(&self, filters: &[Expr]) -> Vec<SeafowlPartition> {
+        let mut partitions = self.partitions.deref().clone();
+
+        if filters.is_empty() {
+            // No qualifiers means we need all partitions
+            return partitions;
+        }
+
+        let mut partition_mask = vec![true; self.partition_count];
+
+        for expr in filters {
+            match PruningPredicate::try_new(expr.clone(), self.schema.clone()) {
+                Ok(predicate) => match predicate.prune(self) {
+                    Ok(expr_mask) => {
+                        partition_mask = partition_mask
+                            .iter()
+                            .zip(expr_mask)
+                            .map(|(&a, b)| a && b)
+                            .collect()
+                    }
+                    Err(error) => {
+                        warn!(
+                            "Failed pruning the partitions for expr {:?}: {}",
+                            expr, error
+                        );
+                    }
+                },
+                Err(error) => {
+                    warn!(
+                        "Failed constructing a pruning predicate for expr {:?}: {}",
+                        expr, error
+                    );
+                }
+            }
+        }
+
+        let mut iter = partition_mask.iter();
+        partitions.retain(|_| *iter.next().unwrap());
+        partitions
+    }
+
+    // Try to convert the vector of min/max scalar values for a column into a generic array reference
+    fn get_values_array(&self, values: Option<&Vec<ScalarValue>>) -> Option<ArrayRef> {
+        match values {
+            Some(stats) => {
+                match ScalarValue::iter_to_array(stats.clone().into_iter()) {
+                    Ok(array) => Some(array),
+                    Err(error) => {
+                        warn!("Failed to convert vector of min/max values {:?} to array: {}", values, error);
+                        None
+                    }
+                }
+            }
+            None => None,
+        }
+    }
```

Comment:
Dunno if this is overkill (feel free to ignore), but every time I see `None => None`, I get an urge to rewrite it

```suggestion
        values.and_then(|stats| {
            ScalarValue::iter_to_array(stats.clone().into_iter())
                .map_err(|e| {
                    warn!(
                        "Failed to convert vector of min/max values {:?} to array: {}",
                        values, e
                    );
                    e
                })
                .ok()
        })
    }
```

Reply:
Yeah, that makes sense.